substrate/frame/revive/src/precompiles/builtin/storage.rs

```diff
@@ -0,0 +1,188 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+	precompiles::{BuiltinAddressMatcher, BuiltinPrecompile, Error, ExtWithInfo},
+	vm::RuntimeCosts,
+	Config, Key, SENTINEL,
+};
+use alloc::vec::Vec;
+use alloy_core::{sol, sol_types::SolValue};
+use core::{marker::PhantomData, num::NonZero};
+use pallet_revive_uapi::StorageFlags;
+use sp_core::hexdisplay::AsBytesRef;
+
+pub struct Storage<T>(PhantomData<T>);
+
+sol! {
+	interface IStorage {
+		/// Clear the value at the given key in the contract storage.
+		///
+		/// # Parameters
+		///
+		/// - `key`: The storage key.
+		///
+		/// # Return
+		///
+		/// Returns the size of the pre-existing value at the specified key if any.
+		function clearStorage(uint32 flags, bool isFixedKey, bytes memory key) external returns (bytes32);
+
+		/// Checks whether there is a value stored under the given key.
+		///
+		/// The key length must not exceed the maximum defined by the contracts module parameter.
+		///
+		/// # Parameters
+		///
+		/// - `key`: The storage key.
+		///
+		/// # Return
+		///
+		/// Returns the size of the pre-existing value at the specified key if any.
+		function containsStorage(uint32 flags, bool isFixedKey, bytes memory key) external returns (bytes32);
```

Comment:
I think if the value size can't be zero this should just return an `uint`, with a return value of 0 indicating there was no value. If there can be keys without values, it should return a tuple `(containedKey, valueLen)`.